crates/red_knot/src/types.rs

```diff
@@ -1,101 +1,327 @@
-use crate::Name;
+use crate::ast_ids::NodeKey;
+use crate::files::FileId;
+use crate::symbols::SymbolId;
+use crate::{FxDashMap, FxIndexSet, Name};
 use ruff_index::{newtype_index, IndexVec};
-use rustc_hash::FxHashSet;
+use rustc_hash::FxHashMap;
 
-#[newtype_index]
-pub(crate) struct TypeId;
+/// unique ID for a type
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+pub(crate) enum Type {
+    /// the dynamic or gradual type: a statically-unknown set of values
+    Any,
+    /// the empty set of values
+    Never,
+    /// unknown type (no annotation)
+    /// equivalent to Any, or to object in strict mode
+    Unknown,
+    /// name is not bound to any value
+    Unbound,
+    /// a specific function
+    Function(FunctionTypeId),
+    /// the set of Python objects with a given class in their __class__'s method resolution order
+    Class(ClassTypeId),
+    Union(UnionTypeId),
+    Intersection(IntersectionTypeId),
+    // TODO protocols, callable types, overloads, generics, type vars
+}
 
-impl TypeId {
-    pub(crate) fn ty(self, env: &TypeEnvironment) -> &Type {
-        env.type_for_id(self)
+impl Type {
+    fn display<'a>(&'a self, store: &'a TypeStore) -> DisplayType<'a> {
+        DisplayType { ty: self, store }
     }
 }
 
-/// Arena holding all known types
+// TODO: currently calling `get_function` et al and holding on to the `FunctionTypeRef` will lock a
+// shard of this dashmap, for as long as you hold the reference. This may be a problem. We could
+// switch to having all the arenas hold Arc, or we could see if we can split up ModuleTypeStore,
+// and/or give it inner mutability and finer-grained internal locking.
 #[derive(Debug, Default)]
-pub(crate) struct TypeEnvironment {
-    types_by_id: IndexVec<TypeId, Type>,
+pub struct TypeStore {
+    modules: FxDashMap<FileId, ModuleTypeStore>,
 }
 
-impl TypeEnvironment {
-    pub(crate) fn add_class(&mut self, name: &str) -> TypeId {
-        let class = Type::Instance {
-            instance_of: ClassType {
-                name: Name::new(name),
-            },
-        };
-        self.types_by_id.push(class)
+impl TypeStore {
+    pub fn remove_module(&mut self, file_id: &FileId) {
+        self.modules.remove(file_id);
     }
 
-    pub(crate) fn add_function(&mut self, name: &str) -> TypeId {
-        let function = Type::Function(FunctionType {
-            name: Name::new(name),
-        });
-        self.types_by_id.push(function)
+    fn add_or_get_module(&mut self, file_id: FileId) -> ModuleStoreRefMut {
+        self.modules
+            .entry(file_id)
+            .or_insert_with(|| ModuleTypeStore::new(file_id))
+    }
+
+    fn get_module(&self, file_id: FileId) -> ModuleStoreRef {
+        self.modules.get(&file_id).expect("module should exist")
+    }
+
+    fn add_function(&mut self, file_id: FileId, name: &str) -> Type {
+        self.add_or_get_module(file_id).add_function(name)
+    }
+
+    fn add_class(&mut self, file_id: FileId, name: &str) -> Type {
```

Comment:
Should this return a `ClassId` similar to `get_class` that returns a `ClassTypeRef`?

Reply:
Yeah, that might actually make some things more convenient, I'll try it out and see how it works out.

Reply:
Yeah this was a good call, done and pushed.